kotlinx-coroutines-core/jvm/src/CoroutineInterruptible.kt

```diff
@@ -0,0 +1,152 @@
+        // Fast path: no Job in this context
+        val threadState = oldState ?: return
+        // Slow path
+        threadState.clearInterrupt()
+    }
+
+    /**
+     * Holds the state of executions for interruption.
+     */
+    @InternalCoroutinesApi
+    class ThreadState {
+        fun initInterrupt(job: Job) {
+            initInvokeOnCancel(job)
+            initThread()
+        }
+
+        fun clearInterrupt() {
+            state.loop { s ->
+                when {
+                    s is Working -> {
+                        if (state.compareAndSet(s, Finish)) {
+                            s.cancelHandle?.let { it.dispose() } // no more watching
+                            return
+                        }
+                    }
+                    s === Interrupting -> Thread.yield() // eases the thread
+                    s === Interrupted -> { Thread.interrupted(); return } // no interrupt leak
+                    s === Init || s === Finish -> throw IllegalStateException("impossible state")
+                    else -> throw IllegalStateException("unknown state")
+                }
+            }
+        }
+
+        private fun initInvokeOnCancel(job: Job) {
+            // watches the job for cancellation
+            val cancelHandle =
+                    job.invokeOnCompletion(onCancelling = true, invokeImmediately = true, handler = CancelHandler())
+            // remembers the cancel handle or drops it
+            state.loop { s ->
+                when {
+                    s === Init -> if (state.compareAndSet(s, Working(null, cancelHandle))) return
+                    s is Working -> if (state.compareAndSet(s, Working(s.thread, cancelHandle))) return
+                    s === Finish -> { cancelHandle.dispose(); return } // no more watching needed
+                    s === Interrupting || s === Interrupted -> return
+                    else -> throw IllegalStateException("unknown state")
+                }
+            }
+        }
+
+        private fun initThread() {
+            val thread = Thread.currentThread()
+            state.loop { s ->
+                when {
+                    s === Init -> if (state.compareAndSet(s, Working(thread, null))) return
+                    s is Working -> if (state.compareAndSet(s, Working(thread, s.cancelHandle))) return
+                    s === Interrupted -> { thread.interrupt(); return } // interrupted before the thread is set
+                    s === Finish || s === Interrupting  -> throw IllegalStateException("impossible state")
+                    else -> throw IllegalStateException("unknown state")
+                }
+            }
+        }
+
+        private inner class CancelHandler : CompletionHandler {
+            override fun invoke(cause: Throwable?) {
+                state.loop { s ->
+                    when {
+                        s === Init || (s is Working && s.thread === null) -> {
```

Comment:
If you swap initThread and initInvokeOnCancel then this can't happen, right?

Reply:
seems a great idea, I will check it